README.md
# extrafont

The extrafont package makes it easier to use fonts other than the basic PostScript fonts that R uses.
Fonts that are imported into extrafont can be used with PDF or PostScript output files. On Windows, extrafont will also make system fonts available for bitmap output.

There are two hurdles for using fonts in PDF (or Postscript) output files:

* Making R aware of the font and the dimensions of the characters.
* Embedding the fonts in the PDF file so that the PDF can be displayed properly on a device that doesn't have the font. This is usually needed if you want to print the PDF file or share it with others.

The extrafont package makes both of these things easier.

Presently it allows the use of TrueType fonts with R, and installation of special font packages.
Support for other kinds of fonts will be added in the future.
It has been tested on Mac OS X 10.7 and Ubuntu Linux 12.04 and Windows XP.


The instructions below are written for PDF files, although the information also applies to PostScript files.

If you want to use the TeX Computer Modern fonts in PDF files, also see the [fontcm](https://github.com/wch/fontcm) package.

# Using extrafont

## Requirements

You must have Ghostscript installed on your system for embedding fonts into PDF files.

Extrafont requires the **[extrafontdb](https://github.com/wch/extrafontdb)** package to be installed.
extrafontdb contains the font database, while this package contains the code to install fonts and register them in the database.

It also requires the **[Rttf2pt1](https://github.com/wch/Rttf2pt1)** package to be installed.
Rttf2pt1 contains the ttf2pt1 program which is used to read and manipulate TrueType fonts.
It is in a separate package for licensing reasons.

Install extrafont from CRAN will automatically install extrafontdb and Rttf2pt1:

```R
install.packages('extrafont')
library(extrafont)
```


To use extrafont in making graphs, you'll need to do the following:

* Import fonts into the extrafont database. (Needs to be done once)
* Register the fonts from the extrafont database with R's PDF (or PostScript) output device. (Needs to be done once per R session)
* Create the graphics that use the fonts.
* Embed the fonts into the PDF file. (Needs to be done for each file)

## Import fonts into the extrafont database

First, import the fonts installed on the system.
(This only works with TrueType fonts right now.)

```R
font_import()
# This tries to autodetect the directory containing the TrueType fonts.
# If it fails on your system, please let me know.
```

This does the following:

* Finds the fonts on your system.
* Extracts the FontName (like ArialNarrow-BoldItalic).
* Extracts/converts a PostScript .afm file for each font. This file contains the *font metrics*, which are the rectangular dimensions of each character that are needed for placement of the characters. These are not the *glyphs*, which the curves defining the visual shape of each character. The glyphs are only in the .ttf file.
* Scan all the resulting .afm files, and save a table with information about them. This table will be used when making plots with R.
* Creates a file `Fontmap`, which contains the mapping from FontName to the .ttf file. This is required by Ghostscript for embedding fonts.


You can view the resulting table of font information with:

```R
# Vector of font family names
fonts()

# Show entire table
fonttable()
```

If you install new fonts on your computer, you'll have to run `font_import()` again.

## Register the fonts with the PDF output device

The next step is to register the fonts in the afm table with R's PDF (or PostScript) output device.
This is needed to create PDF files with the fonts.
As of extrafont version 0.13, this must be run only in the first session when you import your fonts.
In sessions started after the fonts have been imported, simply loading the package with `library(extrafont)` this step isn't necessary, since it will automatically register the fonts with R.

```R
# Only necessary in session where you ran font_import()
loadfonts()
# For PostScript output, use loadfonts(device="postscript")
# Suppress output with loadfonts(quiet=TRUE)
```


## Create figures with the fonts


Here's an example of PDFs made with base graphics and with ggplot2.
These examples use the font Impact, which should be available on Windows and Mac.
(Use `fonts()` to see what fonts are available on your system)

```R
pdf("font_plot.pdf", family="Impact", width=4, height=4)
plot(mtcars$mpg, mtcars$wt, 
     main = "Fuel Efficiency of 32 Cars",
     xlab = "Weight (x1000 lb)",
     ylab = "Miles per Gallon")
dev.off()


library(ggplot2)
p <- ggplot(mtcars, aes(x=wt, y=mpg)) + geom_point() +
  ggtitle("Fuel Efficiency of 32 Cars") +
  xlab("Weight (x1000 lb)") + ylab("Miles per Gallon") +
  theme(text=element_text(size=16, family="Impact"))

ggsave("font_ggplot.pdf", plot=p,  width=4, height=4)
```

The first time you use a font, it may throw some warnings about unknown characters.
This should be harmless, but if it causes any problems, please report them.


## Embed the fonts

After you create a PDF output file, you should *embed* the fonts into the file.
There are 14 PostScript *base fonts* never need to be embedded, because they are included with every PDF/PostScript renderer.
All other fonts should be embedded into the PDF files.


First, if you are running Windows, you may need to tell it where the Ghostscript program is, for embedding fonts. (See Windows installation notes below.)

```R
# Needed only on Windows - run once per R session
# Adjust the path to match your installation of Ghostscript
Sys.setenv(R_GSCMD = "C:/Program Files/gs/gs9.05/bin/gswin32c.exe")
```


As the name suggests, `embed_fonts()` will embed the fonts:

```R
embed_fonts("font_plot.pdf", outfile="font_plot_embed.pdf")
embed_fonts("font_ggplot.pdf", outfile="font_ggplot_embed.pdf")
# If outfile is not specified, it will overwrite the original file
```

To check if the fonts have been properly embedded, open each of the PDF files with Adobe Reader, and go to File->Properties->Fonts.
If a font is embedded, it will say "Embedded Subset" by the font's name; otherwise it will say nothing next to the name.

With Adobe Reader, if a font is not embedded, it will be substituted by another font.
This provides a way to see what your PDF will look like on printer or computer that doesn't have the font installed.
Other PDF viewers may behave differently.
For example, the Preview application on Mac OS X will automatically use system fonts to display non-embedded fonts -- this makes it impossible to tell whether the font is embedded in the PDF.

On Linux you can also use evince (the default PDF viewer) to view embedded fonts.
Open the file and go to File->Properties->Fonts.
If a font is embedded, it will say "Embedded subset"; otherwise it will say "Not embedded".

If you are putting multiple PDF figures into a single document, it is more space-efficient to _not_ embed fonts in each figure, but instead embed the font in the final PDF document.

## Windows bitmap output

extrafont also makes it easier to use fonts in Windows for on-screen or bitmap output.

```R
# Register fonts for Windows bitmap output
loadfonts(device="win")

ggplot(mtcars, aes(x=wt, y=mpg)) + geom_point() +
  ggtitle("Title text goes here") +
  theme(plot.title = element_text(size = 16, family="Georgia", face="italic"))

ggsave("fonttest-win.png")
```

Since the output is a bitmap file, there's no need to embed the fonts.

# Font packages

Extrafont supports _font packages_, which contain fonts that are packaged in a particular way so that they can be imported into extrafont.
These fonts are installed as R packages; they are not installed for the computer operating system.
Fonts that are installed this way will be available only for PDF or PostScript output.
They will not be available for on-screen or bitmap output, which requires that the font be installed for operating system, not just with R and extrafont.

Presently extrafont supports only font packages with PostScript Type 1 fonts.

See the [fontcm](https://github.com/wch/fontcm) package containing Computer Modern fonts for an example.

*****

# Installation notes

## Rttf2pt1

The source code for the utility program `ttf2pt1` is in the package Rttf2pt1.
CRAN has pre-compiled Windows and Mac OS X binaries.
For other platforms, and when installing from source, it will be compiled on installation, so you need a build environment on your system.


## Windows installation notes

In Windows, you need to make sure that Ghostscript is installed.

In each R session where you embed fonts, you will need to tell R where Ghostscript is installed.
For example, when Ghostscript 9.05 is installed to the default location, running this command will do it (adjust the path for your installation):

```R
Sys.setenv(R_GSCMD="C:/Program Files/gs/gs9.05/bin/gswin32c.exe")
```

## Resetting the font database

To reset the extrafont database, reinstall the extrafontdb package:

```R
install.packages("extrafontdb")
```

NEWS
extrafont 0.19
----------------------------------------------------------------

* Fixed #94: the regular expression for finding .pfa and .pfb files was
  incorrect. (#95)

* Added `/usr/local/share/fonts/` to set of search paths on Linux. (#96)

extrafont 0.18
----------------------------------------------------------------

* Extrafont now supports .pfa files in addition to .pfb files. (thanks to Paul
  Murrell)

* `loadfonts()` now has a new default device, `"all"`, which loads fonts for
  all devices. (Thanks to Jeff Arnold)

* Properly handles the encoding of the path of the executable of Rttf2pt1.
  (Thanks to Gabriel Magno)

extrafont 0.17
----------------------------------------------------------------

* Change \donttest{} entries to \dontrun{}.

* Suppress spurious warnings from readLines() in afm_get_info().

* Remove "Character" entry in AFM files because grDevices doesn't understand it.
  (Thanks to gvfarns)

extrafont 0.16
----------------------------------------------------------------

* When importing fonts, don't create duplicate entries. (Fixes #31. Thanks to
  Dylan Schwilk)

extrafont 0.15
----------------------------------------------------------------

* Fix handling of filenames with spaces in R <= 2.15.2.

extrafont 0.14
----------------------------------------------------------------

* Workaround for a bug in ghostscript: if the fontmap path includes the string
  "Resources", it gives an error. Now, in this situation, a temporary symlink
  is made to the fontmap path. (Thanks to Kohske Takahashi)

* `loadfonts(quiet=TRUE)` no longer displays warning messages.

extrafont 0.13
----------------------------------------------------------------

* embed_fonts() now can handle filenames with spaces.

* A new function choose_fonts() has been added, which allows the user to
  specify a set of fonts, and it will return the first one available on the
  current system. (Thanks to Jeffrey Arnold)

* Imported fonts are now automatically registered when extrafont is attached.
  (Thanks to Jeffrey Arnold)

extrafont 0.12
----------------------------------------------------------------

* One of the Linux font search paths has been changed to /usr/share/fonts, so
  that it works with Arch and Gentoo.

* extrafont now searches in the user font directory, ~/.fonts on Linux
  and BSD, and ~/Library/Fonts on Mac OS X.

extrafont 0.11
----------------------------------------------------------------

* extrafont now ignores case when searching for font files.

* All metrics are imported from Truetype fonts. Previously, only the
  metrics for Latin1 were imported. This fixes problems where hyphens
  and minus signs were not allocated enough space. (Thanks to Brian
  Kriegler and Erik Iverson for their investigation of the problem.
  Fixes #11)

* The AFM (font metrics) files are now gzipped.

extrafont 0.10
----------------------------------------------------------------

* A default Truetype font search path for FreeBSD has been added.
  (Thanks to Rainer Hurling)

* extrafont now also makes it easy to use fonts for bitmap output on
  Windows. Use `loadfonts("win")` to register the fonts in Windows.

* loadfonts() now has a `quiet` option to suppress output.


R/load.r
#' Reads the fonttable database and registers those fonts with R
#'
#' This registers fonts so that they can be used with the pdf, postscript, or
#' Windows bitmap output device. It must be run once in each R session.
#'
#' @param device The output device. If \code{"all"}, then it will load
#'   \code{"pdf"}, \code{"postscript"}, and \code{"win"} (if on Windows).
#' @param quiet If \code{FALSE}, print a status message as each font is
#'   registered. If \code{TRUE}, don't print.
#'
#' @return A named list with up to three elements, one for each device for which
#'   fonts were loaded. Each device element is a named list, with an element for
#'   each family that was the function attempted to register with that device.
#'   The value is \code{NULL} if the function did not register the font family
#'   due to problems or because the font family was already registered. If value
#'   is the return value of \code{windowsFonts} for \code{"win"},
#'   \code{\link{postscriptFonts}} for \code{"postscript"}, and
#'   \code{\link{pdfFonts}} for \code{"pdf"}.
#'
#' @seealso \code{\link{embed_fonts}}, #ifdef windows
#'   \code{\link[grDevices]{windowsFont}}, \code{\link[grDevices]{windowsFonts}},
#'   #endif \code{\link{postscriptFonts}}, \code{\link{pdfFonts}},
#'   \code{\link{Type1Font}}.
#' @import grDevices
#' @export
loadfonts <- function(device = c("all", "pdf", "postscript", "win"),
                      quiet = FALSE) {

  device <- match.arg(device)
  if (device == "all") {
    if (.Platform$OS.type == "windows") {
      device <- c("pdf", "postscript", "win")
    } else {
      device <- c("pdf", "postscript")
    }
  }
  ret <- list()
  if ("win" %in% device) {
    ret[["win"]] <- loadfonts_win(quiet = quiet)
  }
  if ("pdf" %in% device) {
    ret[["pdf"]] <- loadfonts_pdf_ps(pdf = TRUE, quiet = quiet)
  }
  if ("postscript" %in% device) {
    ret[["postscript"]] <- loadfonts_pdf_ps(pdf = FALSE, quiet = quiet)
  }
  invisible(ret)
}

register_family_win <- function(family, quiet = FALSE, cfonts = character()) {
  ffname <- "windowsFonts"
  fontfunc <- windowsFonts
  # Now we can register the font with R, with something like this:
  # windowsFonts("Arial" = windowsFont("Arial"))
  if (family %in% cfonts) {
    if (!quiet) {
      message(family, " already registered with ", ffname, "().")
    }
    return(NULL)
  }
  if (!quiet) {
    message("Registering font with R using ", ffname, "(): ", family)
  }
  # Since 'family' is a string containing the name of the argument, we
  # need to use do.call
  args <- list()
  args[[family]] <- windowsFont(family)
  do.call(fontfunc, args)
}

loadfonts_win <- function(quiet = FALSE) {
  ffname <- "windowsFonts"
  if (.Platform$OS.type != "windows") {
    warning("OS is not Windows. No fonts registered with ", ffname, "().")
    return(NULL)
  }
  fontdata <- fonttable()
  # remove empty FamilyNames
  fontdata <- fontdata[fontdata$FamilyName != "", , drop = FALSE]
  cfonts <- names(windowsFonts())
  families <- unique(fontdata$FamilyName)
  lapply(families, register_family_win, cfonts = cfonts, quiet = quiet)
}

register_family_afm <- function(family, fd, pdf = TRUE, quiet = FALSE,
                                cfonts = character()) {
  if (pdf) {
    ffname <- "pdfFont"
    fontfunc <- pdfFonts
  } else {
    ffname <- "postscriptFont"
    fontfunc <- postscriptFonts
  }

  if (family %in% cfonts) {
    if (!quiet) {
      message(family, " already registered with ", ffname, "().")
    }
    return(NULL)
  }

  regular     <- fd$afmfile[!fd$Bold & !fd$Italic]
  bold        <- fd$afmfile[ fd$Bold & !fd$Italic]
  italic      <- fd$afmfile[!fd$Bold &  fd$Italic]
  bolditalic  <- fd$afmfile[ fd$Bold &  fd$Italic]

  # There should be >1 entry for a given weight of a font only for weird
  # fonts like Apple Braille. If found, skip this iteration of the loop.
  if (length(regular) > 1  ||  length(bold)       > 1  ||
      length(italic)  > 1  ||  length(bolditalic) > 1) {
    if (!quiet) {
      message("More than one version of regular/bold/italic found for ",
              family, ". Skipping setup for this font.")
    }
    return(NULL)
  }

  # There should be a regular entry for most every font. Exceptions
  # include Brush Script MT.
  if (length(regular) == 0) {
    if (!quiet) {
      message("No regular (non-bold, non-italic) version of ", family,
              ". Skipping setup for this font.")
    }
    return(NULL)
  }

  # If there aren't bold/italic entries, inherit the afm info from regular
  if (length(bold)       == 0) bold       <- regular
  if (length(italic)     == 0) italic     <- regular
  if (length(bolditalic) == 0) bolditalic <- bold

  # If there's an afmsymfile entry, use that as the symbol font
  # Also check that all in this family have the same afmsymfile entry
  if (!is.na(fd$afmsymfile[1]) && fd$afmsymfile[1] != ""  &&
      all(fd$afmsymfile[1] == fd$afmsymfile)) {
    symbol <- fd$afmsymfile[1]
  } else {
    symbol <- NULL
  }

  # Now we can register the font with R, with something like this:
  # pdfFonts("Arial" = Type1Font("Arial",
  #   file.path(afmpath, c("Arial.afm", "Arial Bold.afm",
  #                        "Arial Italic.afm", "Arial Italic.afm"))))
  if (!quiet) {
    message("Registering font with R using ", ffname, "(): ", family)
  }

  # Since 'family' is a string containing the name of the argument, we
  # need to use do.call
  args <- list()
  args[[family]] <-
    Type1Font(family, metrics = file.path(metrics_path(),
                                          c(regular, bold, italic,
                                            bolditalic, symbol)))
  do.call(fontfunc, args)
}

loadfonts_pdf_ps <- function(pdf = TRUE, quiet = FALSE) {
  if (pdf) {
    # Get names of fonts that are already registered
    cfonts <- names(pdfFonts())
  } else {
    cfonts <- names(postscriptFonts())
  }
  fontdata <- fonttable()
  # remove empty FamilyNames
  fontdata <- fontdata[fontdata$FamilyName != "", , drop = FALSE]
  # split fontdata into list of family data frames
  family_data <- split(fontdata, fontdata$FamilyName)
  mapply(register_family_afm, names(family_data), family_data,
         MoreArgs = list(pdf = pdf, quiet = quiet, cfonts = cfonts),
         SIMPLIFY = FALSE, USE.NAMES = TRUE)
}

#' Embeds fonts that are listed in the local Fontmap
#'
#' @param file Name of input file.
#' @param outfile Name of the output file (with fonts embedded). (Default is same as input file)
#' @param format File format. (see \code{?embedFonts})
#' @param options Other arguments passed to \code{embedFonts}.
#'
#' @examples
#'
#' \dontrun{
#' loadfonts()
#' pdf('fonttest.pdf')
#' library(ggplot2)
#'
#' p <- ggplot(mtcars, aes(x=wt, y=mpg)) + geom_point()
#'
#' # Run only the code below that is appropriate for your system
#' # On Mac and Windows, Impact should be available
#' p + opts(axis.title.x=theme_text(size=16, family="Impact", colour="red"))
#'
#' # On Linux, Purisa may be available
#' p + opts(axis.title.x=theme_text(size=16, family="Purisa", colour="red"))
#' dev.off()
#'
#' embed_fonts('fonttest.pdf', outfile='fonttest-embed.pdf')
#' }
#'
#' @seealso \code{\link{loadfonts}}
#' @export
embed_fonts <- function(file, format, outfile = file, options = "") {

  # This code block is needed in R <= 2.15.2 because of a bug in embedFonts()
  if (getRversion() <= numeric_version("2.15.2")) {
    # This type detection code is necessary because of a bug in embedFonts where
    # it does not correctly detect file type when there are space in the filename.
    # https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=15149
    if (missing(format)) {
      suffix <- gsub(".+[.]", "", file)
      format <- switch(suffix, ps = , eps = "pswrite", pdf = "pdfwrite")
    }

    # To handle spaces, the input file can have quotes, but the output file
    # shouldn't for some reason. We need to force evaluation here; if not, then
    # lazy evaluation will result in outfile having quotes because of the change
    # to file, below.
    force(outfile)

    # Quote filenames so that spaces will work
    file <- shQuote(file)
  }

  fontmap <- fixpath_os(fontmap_path())

  # This is a hack to work around a bug in gs, as of version 9.05. When the
  # fontmap path contains "Resources", it causes a confusing error about
  # GenericResourceDir. On Macs, the default installation directory for R
  # packages contains "Resources" in the path, so this problem is common.
  # The workaround is to create a symlink to the fontmap path, which doesn't
  # contain "Resources".
  if (grepl("^darwin", R.version$os) && grepl("Resources", fontmap)) {
    tmpdir <- tempfile()
    file.symlink(fontmap, tmpdir)
    on.exit(file.remove(tmpdir))
    fontmap <- tmpdir
  }

  embedFonts(file = file, format = format, outfile = outfile,
    options = paste(
      paste("-I", shQuote(fontmap), sep = ""),
      options))
}
